xarray/core/variable.py

```diff
@@ -272,9 +272,19 @@ def data(self, data):
         self._data = data
 
     def _data_cached(self):
-        if not isinstance(self._data, (np.ndarray, PandasIndexAdapter)):
-            self._data = np.asarray(self._data)
-        return self._data
+        """Load data into memory and return it.
+        Do not cache dask arrays automatically; that should
+        require an explicit load() call.
+        """
+        if isinstance(self._data, (np.ndarray, PandasIndexAdapter)):
+            new_data = self._data
+        else:
+            new_data = np.asarray(self._data)
```

Comment:
Can we make a separate method for the logic above? e.g.,
```
def _data_cast(self):
    if isinstance(self._data, (np.ndarray, PandasIndexAdapter)):
        return self._data
    else:
        return np.asarray(self._data)
```

Then both `Variable` and `IndexVariable` can share the caching logic in `_data_cached` and only reimplement `_data_cast`.